File: AcmeHrApp/backend/graph_ql/mutations.py
from graphene import Field, String, ObjectType, Int
from .all_type_defs.hr_mutation_message import HrMutationType, AddDepartmentType
from .all_mutations.hr_manager_mutations.add_employee import add_employee_mutation
from .all_mutations.hr_manager_mutations.edit_employee import edit_employee_mutation
from .all_mutations.hr_manager_mutations.edit_employee_role import edit_employee_role_mutation
from .all_mutations.hr_manager_mutations.edit_employee_department import edit_employee_department_mutation
from .all_mutations.hr_manager_mutations.edit_employee_salary import edit_employee_salary_mutation
from .all_mutations.hr_manager_mutations.add_department import add_department_mutation
from .all_mutations.hr_manager_mutations.add_title import add_title_mutation


class Mutation(ObjectType):
    add_employee = Field(HrMutationType, first_name=String(
        required=True),
        last_name=String(required=True),
        gender=String(required=True),
        birth_date=String(required=True),
        emp_num=String(required=True),
        role=String(required=True),
        department=String(required=True),
        hire_date=String(required=True),
        salary=String(required=True))
    edit_employee = Field(HrMutationType, first_name=String(
        required=True),
        last_name=String(required=True),
        gender=String(required=True),
        birth_date=String(required=True),
        emp_num=String(required=True),
        original_dept_no=String(required=True),
        original_role=String(required=True),
        role=String(required=True),
        department=String(required=True),
        hire_date=String(required=True),
        current_date=String(required=True))
    edit_employee_role = Field(HrMutationType, emp_num=Int(
        required=True),
        department_no=String(required=True),
        original_role=String(required=True),
        new_role=String(required=True),
        new_role_start_date=String(required=True))
    edit_employee_department = Field(HrMutationType, emp_num=Int(
        required=True),
        new_department_no=String(required=True),
        new_department_start_date=String(required=True))
    edit_employee_salary = Field(HrMutationType, emp_num=Int(
        required=True),
        new_salary=String(required=True),
        new_salary_start_date=String(required=True))
    add_department = Field(AddDepartmentType, dept_no=String(
        required=True),
        dept_name=String(required=True),
        managers_emp_num=String(required=True),
        start_date=String(required=True))
    add_title = Field(HrMutationType, title_name=String(
        required=True))


    def resolve_add_employee(root, info, first_name, last_name, gender, birth_date, emp_num, role, department, hire_date, salary):
        response = add_employee_mutation(
            first_name, last_name, gender, birth_date, emp_num, role, department, hire_date, salary)
        return {'message': response}

    def resolve_edit_employee(root, info, first_name, last_name, gender, birth_date, emp_num, original_dept_no, original_role, role, department, hire_date, current_date):
        response = edit_employee_mutation(first_name, last_name, gender, birth_date, emp_num, role, department, hire_date, original_dept_no, original_role, current_date)
        return {'message': response}
    
    def resolve_edit_employee_role(root, info, emp_num, department_no, original_role, new_role, new_role_start_date):
        response = edit_employee_role_mutation(emp_num, department_no, original_role, new_role, new_role_start_date)
        return {'message': response}

    def resolve_edit_employee_department(root, info, emp_num, new_department_no, new_department_start_date):
        response = edit_employee_department_mutation(emp_num, new_department_no, new_department_start_date)
        return {'message': response}

    def resolve_edit_employee_salary(root, info, emp_num, new_salary, new_salary_start_date):
        response = edit_employee_salary_mutation(emp_num, new_salary, new_salary_start_date)
        return {'message': response}

    def resolve_add_department(root, info, dept_no, dept_name, managers_emp_num, start_date):
        response = add_department_mutation(managers_emp_num, dept_no, dept_name, start_date)
        return response

    def resolve_add_title(root, info, title_name):
        response = add_title_mutation(title_name)
        return {'message': response}

File: frontend/src/graphql/mutations.js
import { gql } from "@apollo/client";

const ADD_EMPLOYEE_MUTATION = `
mutation addEmployee(
  $firstName: String!
  $lastName: String!
  $gender: String!
  $birthDate: String!
  $empNum: String!
  $role: String!
  $department: String!
  $hireDate: String!
  $salary: String!
) {
  addEmployee(
    firstName: $firstName
    lastName: $lastName
    gender: $gender
    birthDate: $birthDate
    empNum: $empNum
    role: $role
    department: $department
    hireDate: $hireDate
    salary: $salary
  ) {
    message
  }
}
`;

const ADD_EMPLOYEE = gql`
  ${ADD_EMPLOYEE_MUTATION}
`;

const EDIT_EMPLOYEE_MUTATION = `
mutation editEmployee(
  $firstName: String!
  $lastName: String!
  $gender: String!
  $birthDate: String!
  $empNum: String!
  $originalDeptNo: String!
  $originalRole: String!
  $role: String!
  $department: String!
  $hireDate: String!
  $currentDate: String!
) {
  editEmployee(
    firstName: $firstName
    lastName: $lastName
    gender: $gender
    birthDate: $birthDate
    empNum: $empNum
    originalDeptNo: $originalDeptNo
    originalRole: $originalRole
    role: $role
    department: $department
    hireDate: $hireDate
    currentDate: $currentDate
  ) {
    message
  }
}
`;

const EDIT_EMPLOYEE = gql`
  ${EDIT_EMPLOYEE_MUTATION}
`;

const EDIT_EMPLOYEE_ROLE_MUTATION = `
mutation editEmployeeRole($empNum: Int!, $departmentNo: String!, $originalRole: String!, $newRole: String!, $newRoleStartDate: String!) {
  editEmployeeRole(
    empNum: $empNum
    departmentNo: $departmentNo
    originalRole: $originalRole
    newRole: $newRole
    newRoleStartDate: $newRoleStartDate
  ) {
    message
  }
}
`;

const EDIT_EMPLOYEE_ROLE = gql`
  ${EDIT_EMPLOYEE_ROLE_MUTATION}
`;

const EDIT_EMPLOYEE_DEPARTMENT_MUTATION = `
mutation editEmployeeDepartment($empNum: Int!, $newDepartmentNo: String!, $newDepartmentStartDate: String!) {
  editEmployeeDepartment(empNum: $empNum, newDepartmentNo: $newDepartmentNo, newDepartmentStartDate: $newDepartmentStartDate) {
    message
  }
}
`;

const EDIT_EMPLOYEE_DEPARTMENT = gql`
  ${EDIT_EMPLOYEE_DEPARTMENT_MUTATION}
`;

const EDIT_EMPLOYEE_SALARY_MUTATION = `
mutation editEmployeeSalary($empNum: Int!, $newSalary: String!, $newSalaryStartDate: String!) {
  editEmployeeSalary(empNum: $empNum, newSalary: $newSalary, newSalaryStartDate: $newSalaryStartDate) {
    message
  }
}
`;

const EDIT_EMPLOYEE_SALARY = gql`
  ${EDIT_EMPLOYEE_SALARY_MUTATION}
`;

const ADD_DEPARTMENT_MUTATION = `
mutation addDepartment($deptNo: String!, $deptName: String!, $managersEmpNum: String!, $startDate: String!) {
  addDepartment(deptNo: $deptNo, deptName: $deptName, managersEmpNum: $managersEmpNum, startDate: $startDate) {
    message
    newManager {
      firstName
      lastName
    }
  }
}
`;

const ADD_DEPARTMENT = gql`
  ${ADD_DEPARTMENT_MUTATION}
`;

const ADD_TITLE_MUTATION = `
mutation addTitle($titleName: String!) {
  addTitle(titleName: $titleName) {
    message
  }
}
`;

const ADD_TITLE = gql`
  ${ADD_TITLE_MUTATION}
`;

export {
  ADD_EMPLOYEE,
  ADD_EMPLOYEE_MUTATION,
  EDIT_EMPLOYEE,
  EDIT_EMPLOYEE_MUTATION,
  EDIT_EMPLOYEE_ROLE,
  EDIT_EMPLOYEE_ROLE_MUTATION,
  EDIT_EMPLOYEE_DEPARTMENT,
  EDIT_EMPLOYEE_DEPARTMENT_MUTATION,
  EDIT_EMPLOYEE_SALARY,
  EDIT_EMPLOYEE_SALARY_MUTATION,
  ADD_DEPARTMENT,
  ADD_DEPARTMENT_MUTATION,
  ADD_TITLE,
  ADD_TITLE_MUTATION,
};
